### PR TITLE
*: check region epoch strictly

### PR DESCRIPTION
**Pull request by overvenus**
## What have you changed? 

Checks region epoch strictly to avoid KeyNotInRegion errors.

A 3 nodes TiKV cluster with merge enabled, after commit merge, TiKV A
tells TiDB with a epoch not match error contains the latest target Region
info, TiDB updates its region cache and sends requests to TiKV B,
and TiKV B has not applied commit merge yet, since the region epoch in
request is higher than TiKV B, the request must be denied due to epoch
not match, so it does not read on a stale snapshot, thus avoid the
KeyNotInRegion error.

## What are the type of the changes? (mandatory)

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested? (mandatory)

Integrations tests

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

Cc https://github.com/pingcap/kvproto/pull/344

### CODE REVIEW

File: src/raftstore/errors.rs [hunk 1]
```diff
@@ -115,7 +115,7 @@ quick_error! {
             description("request timeout")
             display("Timeout {}", msg)
         }
-        StaleEpoch(msg: String, new_regions: Vec<metapb::Region>) {
+        EpochNotMatch(msg: String, new_regions: Vec<metapb::Region>) {
             description("region is stale")
```
Comment (zhangjinpeng1987) [inline src/raftstore/errors.rs]: Description and display also need to be changed.

File: src/raftstore/errors.rs [hunk 2]
```diff
@@ -115,7 +115,7 @@ quick_error! {
             description("request timeout")
             display("Timeout {}", msg)
         }
-        StaleEpoch(msg: String, new_regions: Vec<metapb::Region>) {
+        EpochNotMatch(msg: String, new_regions: Vec<metapb::Region>) {
```
Comment (zhangjinpeng1987) [inline src/raftstore/errors.rs]: s/new_regions/current_regions

File: src/storage/mod.rs [hunk 3]
```diff
@@ -1721,7 +1721,7 @@ impl ErrorHeaderKind {
             ErrorHeaderKind::NotLeader => "not_leader",
             ErrorHeaderKind::RegionNotFound => "region_not_found",
             ErrorHeaderKind::KeyNotInRegion => "key_not_in_region",
-            ErrorHeaderKind::StaleEpoch => "stale_epoch",
+            ErrorHeaderKind::EpochNotMatch => "stale_epoch",
```
Comment (zhangjinpeng1987) [inline src/storage/mod.rs]: s/stale_epoch/epoch_not_match

**Comment by zhangjinpeng1987** [comment]
Please fix conflicts

**Comment by Connor1996** [comment]
Please refer the related kvproto PR

**Comment by overvenus** [comment]
@Connor1996 It is reproduced in test_split_epoch_not_match.

**Comment by overvenus** [comment]
PTAL, thanks!

**Comment by siddontang** [comment]
any update about this and corresponding other PRs?

@overvenus @MyonKeminta

**Comment by overvenus** [comment]
/run-integration-tests tidb=release-2.1 pd=release-2.1 tidb-test=release-2.1

**Pull request merged**